lib/using_restricted_array.rb

```diff
@@ -1,61 +1,117 @@
 
 # Calculates the length of the restricted array. All values are integers.
 # The restricted_array is terminated by 'nil' i.e. array[length] = nil
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: linear or O(n) time complexity,where n is the number of items in the array or the length of thei nput
+# Space complexity: constant or O(1)
 def length(array)
-  raise NotImplementedError
+  k = 0
+  while array[k] != nil
+    k += 1
+  end
+  return k
 end
 
 # Prints each integer values in the array
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: linear or O(n)
+# Space complexity: constant or O(1)
+
 def print_array(array)
-  raise NotImplementedError
+  k = 0
+  while array[k] != nil
+    print array[k]
+    print " "
+    k += 1
+  end
+  puts
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: linear O(n), where n is the number of items in the array
+# Space complexity: constant or O(1)
 def search(array, length, value_to_find)
-  raise NotImplementedError
+  length.times do |k|
+    return true if array[k] == value_to_find
+  end
+  return false
 end
 
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n); to find the max value, the array is iterated through n number of times
+# Space complexity: constant or O(1)
 def find_largest(array, length)
-  raise NotImplementedError
+  max_value = array[0]
+  length.times do |k|
+    if array[k] > max_value
+      max_value = array[k]
+    end
+  end
+  return max_value
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n); to find the minimum value, the array is iterated through n number of times
+# Space complexity: constant or O(1)
 def find_smallest(array, length)
-  raise NotImplementedError
+  min_value = array[0]
+  length.times do |k|
+    if array[k] < min_value
+      min_value = array[k]
+    end
+  end
+  return min_value
 end
 
 # Reverses the values in the integer array in place
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n); an array is gone through 2n times
+# Space complexity: linear or O(n)
 def reverse(array, length)
-  raise NotImplementedError
+  temp = []
+  index = length - 1
+  length.times do |k|
+    temp << array[index]
+    index -= 1
+  end
+  length.times do |j|
+    array[j] = temp[j]
+  end
 end
 
 # For an array sorted in ascending order, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(log2)
+# Space complexity: constant or O(1)
 def binary_search(array, length, value_to_find)
-  raise NotImplementedError
+  sort(array, length)
+  mid_index = (length / 2).round
```

Comment:
Could move this whole block into the else case below. You don't need these variables if the value_to_find isn't in the array.